jwk/refresh.go

```diff
@@ -0,0 +1,403 @@
+		}
+	}
+
+	var doReconfigure bool
+	af.muRegistry.Lock()
+	t, ok := af.registry[url]
+	if ok {
+		if t.httpcl != httpcl {
+			t.httpcl = httpcl
+			doReconfigure = true
+		}
+
+		if t.minRefreshInterval != minRefreshInterval {
+			t.minRefreshInterval = minRefreshInterval
+			doReconfigure = true
+		}
+
+		if t.refreshInterval != nil {
+			if !hasRefreshInterval {
+				t.refreshInterval = nil
+				doReconfigure = true
+			} else if *t.refreshInterval != refreshInterval {
+				*t.refreshInterval = refreshInterval
+				doReconfigure = true
+			}
+		} else {
+			if hasRefreshInterval {
+				t.refreshInterval = &refreshInterval
+				doReconfigure = true
+			}
+		}
+	} else {
+		t = &target{
+			httpcl:             httpcl,
+			minRefreshInterval: minRefreshInterval,
+			url:                url,
+			// This is a placeholder timer so we can call Reset() on it later
+			// Make it sufficiently in the future so that we don't have bogus
+			// events firing
+			timer: time.NewTimer(24 * time.Hour),
+		}
+		if hasRefreshInterval {
+			t.refreshInterval = &refreshInterval
+		}
+
+		// Record this in the registry
+		af.registry[url] = t
+		doReconfigure = true
+	}
+	af.muRegistry.Unlock()
+
+	if doReconfigure {
+		// Tell the backend to reconfigure itself
+		af.configureCh <- struct{}{}
+	}
+}
+
+func (af *AutoRefresh) releaseFetching(url string) {
+	// first delete the entry from the map, then close the channel or
+	// otherwise we may end up getting multiple groutines doing the fetch
+	af.muFetching.Lock()
+	fetchingCh, ok := af.fetching[url]
+	if !ok {
+		// Juuuuuuust in case. But shouldn't happen
+		af.muFetching.Unlock()
+		return
+	}
+	delete(af.fetching, url)
+	close(fetchingCh)
+	af.muFetching.Unlock()
+}
+
+func (af *AutoRefresh) Fetch(ctx context.Context, url string) (*Set, error) {
+	af.muRegistry.RLock()
+	_, ok := af.registry[url]
+	if !ok {
+		return nil, errors.Errorf(`url %s must be configured using "Configure()" first`, url)
+	}
+	af.muRegistry.RUnlock()
+
+	ks, found := af.getCached(url)
+	if found {
+		return ks, nil
+	}
+
+	// To avoid a thundering herd, only one goroutine per url may enter into this
+	// initial fetch phase.
+	af.muFetching.Lock()
+	fetchingCh, fetching := af.fetching[url]
+	// unlock happens in each of the if/else clauses because we need to perform
+	// the channel initialization when there is no channel present
+	if fetching {
+		af.muFetching.Unlock()
+		<-fetchingCh
+	} else {
+		fetchingCh = make(chan struct{})
+		af.fetching[url] = fetchingCh
+		af.muFetching.Unlock()
+
+		// Register a cleanup handler, to make sure we always
+		defer af.releaseFetching(url)
+
+		// The first time around, we need to fetch the keyset
+		if err := af.refresh(ctx, url); err != nil {
+			return nil, errors.Wrapf(err, `failed to fetch resource pointed by %s`, url)
+		}
+	}
+
+	// the cache should now be populated
+	ks, ok = af.getCached(url)
+	if !ok {
+		panic("cache was not populated after explicit refresh")
```

Comment:
That should just be an error. The site could be down. Don't want to start throwing panics in that case. The "main" fetcher will return an error from the refresh() function, but the waiters will panic, I don't think that's good.